brainiak/utils/kronecker_solvers.py

```diff
@@ -0,0 +1,314 @@
+    where L = kron(L[0], L[1] .. L[n-1])
+    and L[i] are the lower triangular matrices
+
+    Arguments
+    ---------
+    L : list of 2-D tensors
+        Each element of the list must be a tensorflow tensor and
+        must be a lower triangular matrix of dimension n_i x n_i
+
+    y : 1-D or 2-D tensor
+        Dimension (n_0*n_1*..n_(m-1)) x p
+
+    Returns
+    -------
+    x : 1-D or 2-D tensor
+        Dimension (n_0*n_1*..n_(m-1)) x p
+
+    """
+    n = len(L)
+    if n == 1:
+        return tf.linalg.triangular_solve(L[0], y, adjoint=True)
+    else:
+        x = y
+        na = L[0].get_shape().as_list()[0]
+        n_list = tf.stack([tf.cast(tf.shape(input=mat)[0], dtype=tf.float64) for mat in L])
+        n_prod = tf.cast(tf.reduce_prod(input_tensor=n_list), dtype=tf.int32)
+        nb = tf.cast(n_prod/na, dtype=tf.int32)
+        col = tf.shape(input=x)[1]
+
+        for i in range(na-1, -1, -1):
+            xt, xinb, xina = tf.split(x, [i*nb, nb, (na-i-1)*nb], 0)
+            t = xinb / L[0][i, i]
+            xinb = tf_solve_upper_triangular_kron(L[1:], t)
+            xt = (xt
+                  - tf.reshape(
+                      tf.tile(tf.transpose(a=tf.slice(L[0], [i, 0], [1, i])),
+                              [1, nb*col]),
+                      [i*nb, col])
+                  * tf.reshape(
+                      tf.tile(tf.reshape(t, [-1, 1]), [i, 1]), [i*nb, col]))
+            x = tf.concat(axis=0, values=[xt, xinb, xina])
+
+        return x
+
+
+def tf_kron_mult(L, x):
+    """ Tensorflow multiply with kronecker product matrix
+    Returs kron(L[0], L[1] ...) * x
+
+    Arguments
+    ---------
+    L : list of 2-D tensors
+        Each element of the list must be a tensorflow tensor and
+        must be a square matrix of dimension n_i x n_i
+
+    x : 1-D or 2-D tensor
+        Dimension (n_0*n_1*..n_(m-1)) x p
+
+    Returns
+    -------
+    y : 1-D or 2-D tensor
+        Dimension (n_0*n_1*..n_(m-1)) x p
+    """
+    n = len(L)
+    if n == 1:
+        return tf.matmul(L[0], x)
+    else:
+        na = L[0].get_shape().as_list()[0]
+        n_list = tf.stack([tf.cast(tf.shape(input=mat)[0], dtype=tf.float64) for mat in L])
+        n_prod = tf.cast(tf.reduce_prod(input_tensor=n_list), dtype=tf.int32)
+        nb = tf.cast(n_prod/na, dtype=tf.int32)
+        col = tf.shape(input=x)[1]
+        xt = tf_kron_mult(
+           L[1:],
+           tf.transpose(a=tf.reshape(tf.transpose(a=x), [-1, nb])))
+        y = tf.zeros_like(x)
+        for i in range(na):
+            ya, yb, yc = tf.split(y, [i*nb, nb, (na-i-1)*nb], 0)
+            yb = tf.reshape(tf.matmul(tf.reshape(xt, [nb*col, na]),
+                                      tf.transpose(a=tf.slice(L[0],
+                                                            [i, 0],
+                                                            [1, na]))),
+                            [nb, col])
+            y = tf.concat(axis=0, values=[ya, yb, yc])
```

Comment:
If I understand correctly, this function differs from the previous two functions in that when you calculate yb, it does not depend on ya and yc. So I wonder whether this repeated splitting and concatenation will unnecessarily claim at least twice as much as memory. I think tf.concat will add new node to the graph. If it does not just create a view of the concatenated sub-matrices, I worry that it will actually claim a memory of the size of y for every iteration of i. Do you know if this is the case? Perhaps an alternative is just to create an empty list before the loop. In every iteration you just create yb and append to the list. At the end you only do one concatenation outside the loop.

Reply:
I could be missing something (this is also Narayanan's code), but in my understanding the `tf.split` call in every iteration is different (since it is a function of `i` the loop index). So on every iteration we make a different split, operate on the middle, and recombine. On every iteration of the loop, `ya` and `yc` should be different, so doing the split and concat outside will do the wrong thing. Here's a slightly less terse (matlab) implementation of what should be the same algo https://github.com/pillowlab/fastASD/blob/master/tools_kron/kronmult.m.